transport/mux/provider.go

```diff
@@ -0,0 +1,80 @@
+package mux
+
+import (
+	"context"
+	"net"
+	"sync"
+
+	"github.com/hashicorp/yamux"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+
+	"github.com/temporalio/s2s-proxy/metrics"
+)
+
+type (
+	// MuxProvider manages the process of opening a connection with connProvider, setting up a yamux Session with sessionFn,
+	// and then reporting that session via setNewTransport. If the session closes, a new one will be created and notified
+	// using setNewTransport.
+	MuxProvider struct {
+		name            string
+		connProvider    connProvider
+		sessionFn       func(net.Conn) (*yamux.Session, error)
+		onDisconnectFn  func()
+		setNewTransport SetTransportCallback
+		metricLabels    []string
+		logger          log.Logger
+		shutDown        context.Context
+		startOnce       sync.Once
+	}
+	SetTransportCallback func(session *yamux.Session, conn net.Conn)
+	// connProvider represents a way to get connections, either as a client or a server. MuxProvider guarantees that
+	// Close is called when the provider exits
+	connProvider interface {
+		GetConnection() (net.Conn, error)
+		Close()
+	}
+)
+
+// isDone checks whether a context is done without blocking. Convenience function.
+func isDone(ctx context.Context) bool {
```

Comment:
nit: I think another way to do this is [`ctx.Err() != nil`](https://pkg.go.dev/context#Context.Err)

> If Done is not yet closed, Err returns nil. If Done is closed, Err returns a non-nil error explaining why: DeadlineExceeded if the context's deadline passed, or Canceled if the context was canceled for some other reason. After Err returns a non-nil error, successive calls to Err return the same error.